source/mqtt5_client.c

```diff
@@ -2018,34 +2017,56 @@ PyObject *aws_py_mqtt5_client_get_stats(PyObject *self, PyObject *args) {
         return NULL;
     }
 
-    /* These must be DECREF'd when function ends */
+    /* These must be DECREF'd when function ends on error */
     PyObject *result = NULL;
+    PyObject *incomplete_operation_count_obj = NULL;
+    PyObject *incomplete_operation_size_obj = NULL;
+    PyObject *unacked_operation_count_obj = NULL;
+    PyObject *unacked_operation_size_obj = NULL;
 
     struct aws_mqtt5_client_operation_statistics stats;
     AWS_ZERO_STRUCT(stats);
 
     aws_mqtt5_client_get_stats(client->native, &stats);
 
-    result = PyObject_CallFunction(
-        get_stats_callback_fn_py,
-        "(KKKK)",
-        /* K */ (unsigned long long)stats.incomplete_operation_count,
-        /* K */ (unsigned long long)stats.incomplete_operation_size,
-        /* K */ (unsigned long long)stats.unacked_operation_count,
-        /* K */ (unsigned long long)stats.unacked_operation_size);
+    result = PyTuple_New(4);
     if (!result) {
-        PyErr_WriteUnraisable(PyErr_Occurred());
         goto done;
     }
 
+    incomplete_operation_count_obj = PyLong_FromUnsignedLongLong((unsigned long long)stats.incomplete_operation_count);
+    if (!incomplete_operation_count_obj) {
+        goto done;
+    }
+    incomplete_operation_size_obj = PyLong_FromUnsignedLongLong((unsigned long long)stats.incomplete_operation_size);
+    if (!incomplete_operation_size_obj) {
+        goto done;
+    }
+    unacked_operation_count_obj = PyLong_FromUnsignedLongLong((unsigned long long)stats.unacked_operation_count);
+    if (!unacked_operation_count_obj) {
+        goto done;
+    }
+    unacked_operation_size_obj = PyLong_FromUnsignedLongLong((unsigned long long)stats.unacked_operation_size);
+    if (!unacked_operation_size_obj) {
+        goto done;
+    }
+
+    PyTuple_SET_ITEM(result, 0, incomplete_operation_count_obj);
+    PyTuple_SET_ITEM(result, 1, incomplete_operation_size_obj);
+    PyTuple_SET_ITEM(result, 2, unacked_operation_count_obj);
+    PyTuple_SET_ITEM(result, 3, unacked_operation_size_obj);
     success = true;
 
 done:
+    if (success) {
+        return result;
+    }
 
     Py_XDECREF(result);
+    Py_XDECREF(incomplete_operation_count_obj);
+    Py_XDECREF(incomplete_operation_size_obj);
+    Py_XDECREF(unacked_operation_count_obj);
+    Py_XDECREF(unacked_operation_size_obj);
```

Comment:
DON'T decref these.

PyTuple_SET_ITEM() "steals" a reference to the number object (adds it to the tuple without altering their refcount, essentially transferring ownership to the tuple)

so with this code, if item 2/4 fails, then `Py_XDECREF(result);` will make the `result` get destroyed, which will in turn decref everything inside it, bringing the refcount of`incomplete_operation_count_obj` to 0, which destroys it, but now the `incomplete_operation_count_obj` variable is pointing to garbage memory, and `Py_XDECREF(incomplete_operation_count_obj)` will be like "this is not null, so decref it" which is use-after-free, which is a segfault in the best case, and undefined behavior in the worst case

Reply:
That sounds bad! I adjusted it to avoid this 👍